File: conceitos de linguagem de programacao/projeto4aGustavo/Q2/ackerman.c
#include <stdio.h>
#include <time.h>

int ackerman(int m, int n) {
    if (m == 0) {
        return n + 1;
    } else if (m > 0 && n == 0) {
        return ackerman(m - 1, 1);
    } else {
        return ackerman(m - 1, ackerman(m, n - 1));
    }
}

int main() {
    int m_values[] = {0, 1, 2, 3};
    int n_values[] = {0, 1, 2, 3};
    int m, n;
    double time_spent;
    FILE *fp;
    fp = fopen("c_ackerman_times.csv", "w");
    fprintf(fp, "m,n,time\n");

    for (int i = 0; i < 4; i++) {
        for (int j = 0; j < 4; j++) {
            m = m_values[i];
            n = n_values[j];
            clock_t begin = clock();
            ackerman(m, n);
            clock_t end = clock();
            time_spent = (double)(end - begin) / CLOCKS_PER_SEC;
            fprintf(fp, "%d,%d,%f\n", m, n, time_spent);
        }
    }
    fclose(fp);
    return 0;
}

File: conceitos de linguagem de programacao/projeto4aGustavo/Q2/Ackerman.java
import java.io.FileWriter;
import java.io.IOException;

public class Ackerman {
    public static int ackerman(int m, int n) {
        if (m == 0) {
            return n + 1;
        } else if (m > 0 && n == 0) {
            return ackerman(m - 1, 1);
        } else {
            return ackerman(m - 1, ackerman(m, n - 1));
        }
    }

    public static void main(String[] args) {
        int[] m_values = {0, 1, 2, 3};
        int[] n_values = {0, 1, 2, 3};

        try (FileWriter writer = new FileWriter("java_ackerman_times.csv")) {
            writer.append("m,n,time\n");
            for (int m : m_values) {
                for (int n : n_values) {
                    long start = System.nanoTime();
                    ackerman(m, n);
                    long end = System.nanoTime();
                    double time = (end - start) / 1e9;
                    writer.append(m + "," + n + "," + time + "\n");
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
